python/cudf/dataframe/dataframe.py

```diff
@@ -1469,36 +1459,75 @@ def merge(self, right, on=None, how='inner', left_on=None, right_on=None,
         if how not in ['left', 'inner', 'outer']:
             raise NotImplementedError('{!r} merge not supported yet'
                                       .format(how))
+
+        # Making sure that the "on" arguments are list of column names
+        if on:
+            on = [on] if isinstance(on, str) else list(on)
+        if left_on:
+            left_on = [left_on] if isinstance(left_on, str) else list(left_on)
+        if right_on:
+            right_on = ([right_on] if isinstance(right_on, str)
+                        else list(right_on))
+
+        lhs = self.copy(deep=False)
+        rhs = right.copy(deep=False)
+
         same_names = set(lhs.columns) & set(rhs.columns)
         if same_names and not (lsuffix or rsuffix):
             raise ValueError('there are overlapping columns but '
                              'lsuffix and rsuffix are not defined')
 
-        def fix_name(name, suffix):
-            if name in same_names:
-                return "{}{}".format(name, suffix)
-            return name
+        tmp_columns = []  # Columns NOT to include in the final result
 
+        # Let's find the column to do the merge on.
+        # Since `cpp_join.join()` doesn't take indexes
+        # we insert indexes as regular columns
         if left_index and right_index:
-            on = lhs.LEFT_RIGHT_INDEX_NAME
-            lhs[on] = lhs.index
-            rhs[on] = rhs.index
-        if on is None and left_on is None and right_on is None:
+            on = [lhs.LEFT_RIGHT_INDEX_NAME]
+            lhs[on[0]] = lhs.index
+            rhs[on[0]] = rhs.index
+        elif on:
+            if left_on or right_on:
+                raise ValueError('Can only pass argument "on" OR "left_on" '
+                                 'and "right_on", not a combination of both.')
+        elif left_index and right_on:
+            if len(right_on) != 1:
+                raise ValueError('right_on should be a single column')
+            on = right_on
+            lhs[on[0]] = lhs.index
```

Comment:
Doesn't this have the potential to destroy data from `lhs`, if `right_on` specifies a column that actually exists in `lhs`?

Reply:
Yes, I now use a new column name